API: Intermittent JWT Validation Failures Due to Missing JWK Set Cache

## Summary

The API container periodically throws `UnknownHostException` when validating JWTs, causing authentication failures for users. The root cause is that Spring Security is using `NoOpCache` for the JWK set, meaning it makes a live network call to the Zitadel JWKS endpoint on every cache expiry. Any brief DNS hiccup inside the Docker container surfaces as an auth error.

## Error

```
Caused by: org.springframework.security.oauth2.jwt.JwtException: An error occurred while attempting to decode the Jwt:
  I/O error on GET request for "https://budget-buddy-zk5ltr.eu1.zitadel.cloud/oauth/v2/keys":
  budget-buddy-zk5ltr.eu1.zitadel.cloud

Caused by: java.net.UnknownHostException: budget-buddy-zk5ltr.eu1.zitadel.cloud
```

Key indicator in the stack trace — Spring's cache abstraction is not configured:

```
at org.springframework.cache.support.NoOpCache.get(NoOpCache.java:76)
at ...NimbusJwtDecoder$JwkSetUriJwtDecoderBuilder$SpringJWKSource.getJWKSet(NimbusJwtDecoder.java:544)
```

## Root Cause

`NimbusJwtDecoder` is constructed without a cache, so Spring falls back to `NoOpCache`. This means the JWKS (public keys used to verify JWTs) are re-fetched from Zitadel frequently. Public keys change only during key rotation, so there is no reason to make network calls more than a few times per day. Any transient DNS failure during a fetch fails the entire JWT validation and returns a 401 to the user.

## Fix

Configure `NimbusJwtDecoder` with a Caffeine-backed Spring cache.

**1. Add dependencies**

```xml
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-cache</artifactId>
</dependency>
<dependency>
    <groupId>com.github.ben-manes.caffeine</groupId>
    <artifactId>caffeine</artifactId>
</dependency>
```

**2. Configure cache TTL in `application.yml`**

```yaml
spring:
  cache:
    type: caffeine
    caffeine:
      spec: "expireAfterWrite=1h"
```

**3. Inject the cache into `NimbusJwtDecoder`**

```java
@Bean
JwtDecoder jwtDecoder(CacheManager cacheManager) {
    return NimbusJwtDecoder.withJwkSetUri(jwkSetUri)
        .cache(cacheManager.getCache("jwks"))
        .build();
}
```

## Expected Outcome

- JWKS is fetched from Zitadel once per hour instead of on every cache expiry cycle
- Transient DNS failures inside the Docker container no longer cause JWT validation errors
- No user-visible 401s from network instability

## Notes

- Spring Boot does not expose `application.properties` knobs for JWK cache TTL — programmatic configuration is required (tracked in [spring-projects/spring-security#8325](https://github.com/spring-projects/spring-security/issues/8325))
- A 1-hour TTL is safe since Zitadel public keys change only during manual key rotation

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -38,7 +38,9 @@ dependencies {
   implementation("org.springframework.boot:spring-boot-starter-oauth2-resource-server")
   implementation("org.springframework.boot:spring-boot-starter-data-jdbc")
   implementation("org.springframework.boot:spring-boot-starter-actuator")
+  implementation("org.springframework.boot:spring-boot-starter-cache")
   implementation("org.springframework.boot:spring-boot-starter-liquibase")
+  implementation("com.github.ben-manes.caffeine:caffeine")
   implementation("org.openapitools:jackson-databind-nullable:${jacksonDatabindNullableVersion}")
   implementation("org.mapstruct:mapstruct:${mapstructVersion}")
 

--- a/src/integrationTest/java/com/budget/buddy/budget_buddy_api/security/oidc/JwtValidationConfigIntegrationTest.java
+++ b/src/integrationTest/java/com/budget/buddy/budget_buddy_api/security/oidc/JwtValidationConfigIntegrationTest.java
@@ -1,9 +1,12 @@
 package com.budget.buddy.budget_buddy_api.security.oidc;
 
 import com.budget.buddy.budget_buddy_api.BaseIntegrationTest;
+import com.budget.buddy.budget_buddy_api.base.config.CacheConfig;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.security.oauth2.server.resource.autoconfigure.servlet.JwkSetUriJwtDecoderBuilderCustomizer;
+import org.springframework.cache.CacheManager;
 import org.springframework.core.env.Environment;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -22,6 +25,12 @@ class JwtValidationConfigIntegrationTest extends BaseIntegrationTest {
   @Autowired
   private Environment environment;
 
+  @Autowired
+  private CacheManager cacheManager;
+
+  @Autowired
+  private JwkSetUriJwtDecoderBuilderCustomizer jwkSetUriJwtDecoderBuilderCustomizer;
+
   @Test
   @DisplayName("should have audience validation configured")
   void shouldHaveAudienceValidationConfigured() {
@@ -38,5 +47,21 @@ class JwtValidationConfigIntegrationTest extends BaseIntegrationTest {
     assertThat(issuerUri)
         .as("issuer-uri must be set for JWT signature and issuer validation")
         .isNotBlank();
+  }
+
+  @Test
+  @DisplayName("should register the jwks cache so NimbusJwtDecoder can avoid NoOpCache")
+  void shouldRegisterJwksCache() {
+    assertThat(cacheManager.getCache(CacheConfig.JWKS))
+        .as("jwks cache must be registered so the JWT decoder can cache the JWKS response")
+        .isNotNull();
+  }
+
+  @Test
+  @DisplayName("should expose a JwkSetUriJwtDecoderBuilderCustomizer that attaches the jwks cache")
+  void shouldExposeJwkSetUriJwtDecoderBuilderCustomizer() {
+    assertThat(jwkSetUriJwtDecoderBuilderCustomizer)
+        .as("the cache customizer bean must be present so Spring Boot applies it when building the NimbusJwtDecoder")
+        .isNotNull();
   }
 }

--- a/src/main/java/com/budget/buddy/budget_buddy_api/base/config/CacheConfig.java
+++ b/src/main/java/com/budget/buddy/budget_buddy_api/base/config/CacheConfig.java
@@ -15,11 +15,7 @@ public class CacheConfig {
   /** Maps {@code (issuer, subject) → localUserId} for the OIDC JIT-provisioning lookup. */
   public static final String LOCAL_USER_IDS = "localUserIds";
 
-  /**
-   * Caches JWKS responses from the OIDC issuer. Without this, {@code NimbusJwtDecoder} falls back
-   * to {@code NoOpCache} and refetches the JWKS on every cycle, surfacing transient DNS failures
-   * as 401s.
-   */
+  /** Caches JWKS responses so {@code NimbusJwtDecoder} doesn't refetch them on every cycle. */
   public static final String JWKS = "jwks";
 
   @Bean

--- a/src/main/java/com/budget/buddy/budget_buddy_api/base/config/CacheConfig.java
+++ b/src/main/java/com/budget/buddy/budget_buddy_api/base/config/CacheConfig.java
@@ -1,8 +1,10 @@
 package com.budget.buddy.budget_buddy_api.base.config;
 
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.time.Duration;
 import org.springframework.cache.CacheManager;
 import org.springframework.cache.annotation.EnableCaching;
-import org.springframework.cache.concurrent.ConcurrentMapCacheManager;
+import org.springframework.cache.caffeine.CaffeineCacheManager;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
@@ -13,8 +15,17 @@ public class CacheConfig {
   /** Maps {@code (issuer, subject) → localUserId} for the OIDC JIT-provisioning lookup. */
   public static final String LOCAL_USER_IDS = "localUserIds";
 
+  /**
+   * Caches JWKS responses from the OIDC issuer. Without this, {@code NimbusJwtDecoder} falls back
+   * to {@code NoOpCache} and refetches the JWKS on every cycle, surfacing transient DNS failures
+   * as 401s.
+   */
+  public static final String JWKS = "jwks";
+
   @Bean
   CacheManager cacheManager() {
-    return new ConcurrentMapCacheManager(LOCAL_USER_IDS);
+    var cacheManager = new CaffeineCacheManager(LOCAL_USER_IDS, JWKS);
+    cacheManager.setCaffeine(Caffeine.newBuilder().expireAfterWrite(Duration.ofHours(1)));
+    return cacheManager;
   }
 }

--- a/src/main/java/com/budget/buddy/budget_buddy_api/security/SecurityConfig.java
+++ b/src/main/java/com/budget/buddy/budget_buddy_api/security/SecurityConfig.java
@@ -47,12 +47,7 @@ public class SecurityConfig {
     return http.build();
   }
 
-  /**
-   * Attaches the {@link CacheConfig#JWKS} cache to the auto-configured {@code NimbusJwtDecoder}
-   * so the JWKS is fetched from the OIDC issuer at most once per cache TTL instead of on every
-   * cycle. Without a cache the decoder falls back to {@code NoOpCache}, surfacing transient DNS
-   * failures inside the container as 401s.
-   */
+  /** Attaches the {@link CacheConfig#JWKS} cache to the auto-configured {@code NimbusJwtDecoder}. */
   @Bean
   JwkSetUriJwtDecoderBuilderCustomizer jwkSetUriJwtDecoderBuilderCustomizer(
       CacheManager cacheManager

--- a/src/main/java/com/budget/buddy/budget_buddy_api/security/SecurityConfig.java
+++ b/src/main/java/com/budget/buddy/budget_buddy_api/security/SecurityConfig.java
@@ -1,7 +1,10 @@
 package com.budget.buddy.budget_buddy_api.security;
 
+import com.budget.buddy.budget_buddy_api.base.config.CacheConfig;
 import com.budget.buddy.budget_buddy_api.security.oidc.OidcUserProvisioningFilter;
 import com.budget.buddy.budget_buddy_api.user.UserService;
+import org.springframework.boot.security.oauth2.server.resource.autoconfigure.servlet.JwkSetUriJwtDecoderBuilderCustomizer;
+import org.springframework.cache.CacheManager;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.security.config.Customizer;
@@ -42,5 +45,18 @@ public class SecurityConfig {
         .formLogin(AbstractHttpConfigurer::disable);
 
     return http.build();
+  }
+
+  /**
+   * Attaches the {@link CacheConfig#JWKS} cache to the auto-configured {@code NimbusJwtDecoder}
+   * so the JWKS is fetched from the OIDC issuer at most once per cache TTL instead of on every
+   * cycle. Without a cache the decoder falls back to {@code NoOpCache}, surfacing transient DNS
+   * failures inside the container as 401s.
+   */
+  @Bean
+  JwkSetUriJwtDecoderBuilderCustomizer jwkSetUriJwtDecoderBuilderCustomizer(
+      CacheManager cacheManager
+  ) {
+    return builder -> builder.cache(cacheManager.getCache(CacheConfig.JWKS));
   }
 }

